frontend/catalyst/third_party/oqc/src/OQCDevice.hpp

```diff
@@ -52,6 +52,9 @@ class OQCDevice final : public Catalyst::Runtime::QuantumDevice {
     Catalyst::Runtime::CacheManager<std::complex<double>> cache_manager{};
     bool tape_recording{false};
     size_t device_shots;
+    
+    std::string seed;
```

Comment:
I think you are right, the device only needs the prng, not the seed itself. 

I carried the seed over to the device basically just for the `has_seed`, but I guess `has_gen` would serve the same purpose.

Reply:
Note that the prng has to go with the device, not the context; otherwise multiple devices (or multiple instances of the same device) would in principle share a prng, and calling the same device again would not repeat the evolution from the initial prng state anymore!

Reply:
From a UI point of view if the seed is in the `@qjit` decorator then the seed has to be in the context. And if the seed is in the `qnode` then it should be in the device. That is my opinion.

Reply:
> calling the same device again would not repeat the evolution from the initial prng state anymore!

I don't follow, what do you mean by this?

I think this should be true.

```python
#pseudo-code

def wrapper():
  return circuit()

@qml.qnode(...)
def circuit(x):
  qml.Hadamard()
  return qml.measure()

a = qjit(seed=...)(wrapper)
b = qjit(seed=...)(wrapper)
for x in range(100):
  assert a() == b()
```

Also:

```python
def wrapper():
  return circuit()

@qml.qnode(...)
def circuit(x):
  qml.Hadamard()
  return qml.measure()

a = qjit(seed=...)(wrapper)
seq0 = [a() for x in range(100)]
seq1 = [a() for x in range(100)]

assert seq0 != seq1
```

Reply:
Right now we implement this behavior (without seeding the shots yet): 

> @dime10 Ah, so
> 
> ```python
> dev = qml.device("lightning.qubit", wires=1, shots=None)
> 
> @qjit(seed="qwerty")
> def workflow():
>     @qml.qnode(dev)
>     def circuit():
>         qml.Hadamard(0)
>         m = measure(0)
>         @cond(m)
>         def cfun0():
>             qml.Hadamard(0)
>         cfun0()
>         return qml.probs()
> 
>     return circuit(), circuit(), circuit(), circuit()
> 
> print(workflow())
> print("#######")
> print(workflow())
> ```
> 
> ```
> (Array([0.4975, 0.5025], dtype=float64), Array([1., 0.], dtype=float64), Array([1., 0.], dtype=float64), Array([1., 0.], dtype=float64))
> #######
> (Array([0.4975, 0.5025], dtype=float64), Array([1., 0.], dtype=float64), Array([1., 0.], dtype=float64), Array([1., 0.], dtype=float64))
> ```
> 
> i.e. the history of both `workflow` (which is the `QJIT`) calls are the same, but within each QJIT call the qnode results are different?

I think this is what your example meant as well? 
First example: different calls to qjits with the same seed (or the same qjit) should produce the same results 
Second example: keep calling the same qjit gives random results

In the current implementation the random number generator object (the `std::mt19937 gen`) is always in the context; the devices only carry pointers to the rng object. The seed isn't actually important as it is just a string. Once a rng object is made from the seed the seed string itself can be forgotten (which I think was David'd original suggestion).

Reply:
A lot of this confusion comes from the fact that I haven't added frontend tests. I will do so asap to clear up the behavior.

Reply:
Also, something to keep in mind, if the generator is shared among multiple devices, then it should be wrapped in a mutex. While we don't use async much, but we should lock shared resources.

Reply:
> Note that the prng has to go with the device, not the context; otherwise multiple devices (or multiple instances of the same device) would in principle share a prng, and calling the same device again would not repeat the evolution from the initial prng state anymore!

I think multiple devices sharing the prng instance (as a weak reference, while the context is owning it) is actually what we want. That way the randomness is different between repeated invocations of the same (or different) quantum functions.

> Also, something to keep in mind, if the generator is shared among multiple devices, then it should be wrapped in a mutex. While we don't use async much, but we should lock shared resources.

Yeah I briefly thought about this and thought we could just disallow using seeding and async together. I didn't consider that devices might internally be threading and needing access to the rng instance from their threads, but in that case I think they can manage their thread-safety internally.

Also happy to lock it though if we want, just thought it would be unnecessary for our purposes.

Reply:
I have added a check in `CompilerOptions` to disallow seeding with async.